### ISSUE TITLE
Cmd+Q does not close Zenmap

### ISSUE THREAD

**Issue by Varunram** [nmap/nmap#658]
Facing the same issue as reported by Michael in the dev-list. Currently running OSX 10.12. Is this limited to 10.12 or is everyone facing the same problem?

**Comment by W0naN0w** [comment]
As far as I know, this is not an OSX 10.12 bug, it happens on previous OS versions... I believe this is due to the Gui library, which may not handle properly the shortcuts (in a cross platform way, as in Qt..). Still, there could be a kind of a runtime check to see on which platform the software is running, thus providing different shortcuts according to the OS in use.
What do you think?
Cheers

**Comment by Varunram** [comment]
>>> there could be a kind of a runtime check to see on which platform the software is running

This or we could hardcode the values for a Darwin like system in main_actions because the default key is control in all other systems. Which of these would be be better?

**Comment by W0naN0w** [comment]
Yes you are right, this looks more straightforward and better architectured. What about a compile-time check? ... Because this is already in the source code. Would that work with the installer? It'd be less computation during runtime. What do you think?

**Comment by Varunram** [comment]
It seems cool but how would you run a compile time check in Python?

**Comment by W0naN0w** [comment]
Obviously not in python, I was thinking about the config file or the makefile, but then I don't think this wohld work if the user install zenmap via thr installer. Still a runtime check is not really expensive :) if needed I can have a look at it this weekend

**Comment by Varunram** [comment]
Yeah, a few options are available. I'm sure Dan would have an idea on how to go about implementing this.

**Comment by Varunram** [comment]
@dmiller-nmap Please suggest a way to go about doing this

**Comment by dmiller-nmap** [comment]
Ok, here's what I found out so far:

* The command-Q and other command key combos send signals/events that are OS X-specific. The interface for registering functions to handle these events is in the `gtkosx_application` library.
* The particular event for quitting seems to be `NSApplicationBlockTermination`. The handler should do any cleanup we want and then return FALSE (that's how it works in C, not sure if the binding works properly for Python).
* Since I don't have an Apple keyboard, I wrote a simple AppleScript to send the appropriate keystrokes:

```applescript
tell application "Zenmap"
  activate
end tell
delay 0.1
tell application "System Events"
  keystroke "q" using command down 
end tell
```

* I attempted to make this work, but I can't get the handler to fire. Here's my patch:

```diff
Index: zenmap/zenmapGUI/App.py
===================================================================
--- zenmap/zenmapGUI/App.py     (revision 36559)
+++ zenmap/zenmapGUI/App.py     (working copy)
@@ -258,13 +258,17 @@
     sys.excepthook = excepthook
 
 
+def shutdown_cleanup():
+    for window in open_windows:
+        window.scan_interface.kill_all_scans()
+
+
 def safe_shutdown(signum, stack):
     """Kills any active scans/tabs and shuts down the application."""
     log.debug("\n\n%s\nSAFE SHUTDOWN!\n%s\n" % ("#" * 30, "#" * 30))
     log.debug("SIGNUM: %s" % signum)
 
-    for window in open_windows:
-        window.scan_interface.kill_all_scans()
+    shutdown_cleanup()
 
     sys.exit(signum)
 
@@ -275,6 +279,19 @@
     signal.signal(signal.SIGTERM, safe_shutdown)
     signal.signal(signal.SIGINT, safe_shutdown)
 
+    frozen = getattr(sys, "frozen", None)
+    osx_app = None
+    if frozen == "macosx_app" or "Zenmap.app" in sys.executable:
+        # A py2app .app bundle.
+        import gtkosx_application
+        osx_app = gtkosx_application.OSXApplication()
+        def osx_safe_shutdown(*args, **kwargs):
+            shutdown_cleanup()
+            # Should termination be blocked? False.
+            return False
+        osx_app.connect('NSApplicationBlockTermination', osx_safe_shutdown)
+        osx_app.ready()
+
     DEVELOPMENT = os.environ.get(APP_NAME.upper() + "_DEVELOPMENT", False)
     if not DEVELOPMENT:
         install_excepthook()
```

**Comment by Varunram** [comment]
Tried a couple of different handlers but none worked. The patch builds fine though. Did you try changing keys via "Modifier keys"?

**Comment by fyodor** [comment]
I am still able to reproduce this on Mac.  We are looking into it.
